viewing_party/party.py

```diff
@@ -1,23 +1,166 @@
-
 def create_movie(title, genre, rating):
-    pass
+    if title != None and genre != None and rating != None:
+        movie_dict = {}
+        movie_dict["title"] = title
+        movie_dict["genre"] = genre
+        movie_dict["rating"] = rating
+    else:
+        return None
+    return movie_dict
+
+
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
+
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+
+def watch_movie(user_data, title):
+    for movie in user_data["watchlist"]:
+        if movie["title"] == title:
+            user_data["watchlist"].remove(movie)
+            user_data = add_to_watched(user_data, movie)
+    return user_data
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    watched_movies_list = []
+    sum = 0
+    for item in user_data["watched"]:
+        watched_movies_list.append(item["rating"])
+    for number in watched_movies_list:
+        sum += number
+    if watched_movies_list:
+        return sum / len(watched_movies_list)
+    else:
+        return 0.0
+
+
+def get_most_watched_genre(user_data):
+    most_watched = {}
+    for movie in user_data["watched"]:
+        most_watched.setdefault(movie["genre"], 0)
+        most_watched[movie["genre"]] += 1
+
+    max_value = 0
+    max_key = None
+
+    for key, value in most_watched.items():
+        if value > max_value:
+            max_value = value
+            max_key = key
+    return max_key
 
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    list_watched = []
+    for movie in user_data["watched"]:
+        list_watched.append(movie)
+
+    friends_watched_set = set()
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched_set.add(movie["title"])
+
+    list_unique = []
+    for movie in list_watched:
+        if movie["title"] not in friends_watched_set:
+            list_unique.append(movie)
+    return list_unique
+
+
+def get_friends_unique_watched(user_data):
+    set_watched = set()
+    for movie in user_data["watched"]:
+        set_watched.add(movie["title"])
+
+    friends_watched_dict = {}
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_watched_dict[movie["title"]] = movie
+
+    list_unique = []
+    for key, value in friends_watched_dict.items():
+        if key not in set_watched:
+            list_unique.append(value)
+
+    return list_unique
+
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):

```

Comment:
Great implementation, though we could've used `get_friends_unique_watched` as a helper function to avoid any redundancy in code logic!